prqlc/prqlc/src/cli/docs_generator.rs

```diff
@@ -0,0 +1,430 @@
+//         }
+//         docs.push_str("</ul>\n");
+//     }
+
+//     for stmt in stmts
+//         .clone()
+//         .into_iter()
+//         .filter(|stmt| matches!(stmt.kind, StmtKind::VarDef(_)))
+//     {
+//         let var_def = stmt.kind.as_var_def().unwrap();
+//         if var_def.kind != VarDefKind::Let {
+//             continue;
+//         }
+
+//         docs.push_str("<section>\n");
+//         docs.push_str(&format!(
+//             "  <h3 id=\"fn-{}\">{}</h3>\n",
+//             var_def.name, var_def.name
+//         ));
+
+//         //if let Some(docComment) = vardef.DocComment {
+//         //    docs.push_str(&format!("  <p>{docComment}</p>\n"));
+//         //}
+
+//         if let Some(expr) = &var_def.value {
+//             match &expr.kind {
+//                 ExprKind::Func(func) => {
+//                     docs.push_str("  <h4 class=\"h6\">Parameters</h4>\n");
+//                     docs.push_str("  <ul>\n");
+//                     for param in &func.params {
+//                         docs.push_str(&format!("    <li><var>{}</var></li>\n", param.name));
+//                     }
+//                     docs.push_str("  </ul>\n");
+
+//                     if let Some(return_ty) = &func.return_ty {
+//                         docs.push_str("  <h4 class=\"h6\">Returns</h4>\n");
+//                         match &return_ty.kind {
+//                             TyKind::Any => docs.push_str("  <p>Any</p>\n"),
+//                             TyKind::Ident(ident) => {
+//                                 docs.push_str(&format!("  <p><code>{}</code></p>\n", ident.name));
+//                             }
+//                             TyKind::Primitive(primitive) => {
+//                                 docs.push_str(&format!("  <p><code>{primitive}</code></p>\n"));
+//                             }
+//                             TyKind::Singleton(literal) => {
+//                                 docs.push_str(&format!("  <p><code>{literal}</code></p>\n"));
+//                             }
+//                             TyKind::Union(vec) => {
+//                                 docs.push_str("  <ul class=\"list-unstyled\">\n");
+//                                 for (_, ty) in vec {
+//                                     docs.push_str(&format!("    <li>{:?}</li>\n", ty.kind));
+//                                 }
+//                                 docs.push_str("  </ul>\n");
+//                             }
+//                             _ => docs.push_str("  <p class=\"text-danger\">Not implemented</p>\n"),
+//                         }
+//                     }
+//                 }
+//                 ExprKind::Pipeline(_) => {
+//                     docs.push_str("  <p>There is a pipeline.</p>\n");
+//                 }
+//                 _ => (),
+//             }
+//         }
+
+//         docs.push_str("</section>\n");
+//     }
+
+//     html.replacen("{{ content }}", &docs, 1)
+// }
```

Comment:
@vanillajonathan we should probably remove this big comment block!

Reply:
Or maybe we should add a `doc-html` command?

Reply:
We could add `--format=html`, but:
- We would want some reason to use HTML rather than Markdown — something that an html markdown generator couldn't give us.
- If there are ways of improving the current one, I think that's more productive than adding more formats at the moment. It's a cool idea at the moment, but not _that_ helpful. For example, how does it handle multi-file projects?

Reply:
HTML is directly usable and easy to use and publish without having any Markdown viewer installed or use any Markdown-to-HTML processor or static site generator.

HTML allows us rich semantics through the use the `<var>` element to define variables, and the `<time>` element for when the documentation was generated.

It does not yet handle multi-file projects. I created a simple implementation which can hopefully be extended in the future if the need should arise.

Reply:
OK. I really would encourage trying to make something that will be useful for users — that's the best way to get feedback on what's working, what else would be helpful, etc.

Adding multi-file projects could be great. There is already a good implementation of something that collects together multiple files in `prqlc collect`; it shouldn't need to collect the files itself. I could see that being the most useful part of something like `prqlc doc` — you have a large project and want to have a single page on what it contains.

> HTML is directly usable and easy to use and publish without having any Markdown viewer installed or use any Markdown-to-HTML processor or static site generator.

But we should try and have our tools be orthogonal — we don't need to make HTML directly unless there's a good reason to not just use a standard converter...

> HTML allows us rich semantics through the use the `<var>` element to define variables, and the `<time>` element for when the documentation was generated.

Yes, true.

But then when we need to make the tool better, we need to update the HTML generator too! Which then makes it harder to make the tool better...

---

If you're looking for other things to do — we very much appreciate your contributions — then improving code coverage and improving error messages would both be really helpful.